api/server/sdk/volume_node_ops.go

```diff
@@ -122,6 +122,56 @@ func (s *VolumeServer) Detach(
 	return &api.SdkVolumeDetachResponse{}, nil
 }
 
+func (s *VolumeServer) ControllerPublish(
+	ctx context.Context,
+	req *api.SdkVolumeControllerPublishRequest,
+) (*api.SdkVolumeControllerPublishResponse, error) {
+	if s.cluster() == nil || s.driver(ctx) == nil {
+		return nil, status.Error(codes.Unavailable, "Resource has not been initialized")
+	}
+
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.Unavailable, "VolumeId not provided.")
+	}
+
+	opts := make(map[string]string)
+	context, err := s.driver(ctx).ControllerPublish(ctx, req.GetVolumeId(), req.GetNodeId(), opts)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"Failed to publish the volume %v: %v", req.GetVolumeId(),
+			err.Error())
+	}
+
+	return &api.SdkVolumeControllerPublishResponse{
+		Context: context,
+	}, nil
+}
+
+func (s *VolumeServer) ControllerUnpublish(
+	ctx context.Context,
+	req *api.SdkVolumeControllerUnpublishRequest,
+) (*api.SdkVolumeControllerUnpublishResponse, error) {
+	if s.cluster() == nil || s.driver(ctx) == nil {
+		return nil, status.Error(codes.Unavailable, "Resource has not been initialized")
+	}
+
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.Unavailable, "VolumeId not provided.")
+	}
+
+	opts := make(map[string]string)
+	err := s.driver(ctx).ControllerUnpublish(ctx, req.GetVolumeId(), req.GetNodeId(), opts)
```

Comment:
Lets call the actual driver's Attach/Detach/Mount/Unmount functions here.

Reply:
What you are implying here is that instead of having a ControlerPublish equivalent on the porx side, just have Attach()/Detach () and in CSI driver's CP/CUP call s.Driver.Attach()/s.Driver.Detach() respectively for windows shares? That though is a workable code, it is not clean code.

Reply:
Hi @adityadani I understand your concern with bringing in new code flow to time tested workflow in sharedv4 code path. But in the long run, I see added value in clearly distinguishing functionality provided by this extension of using CSI ControllerPublish and ControllerUnpublish. Our convoluted way of using proxy objects and filter in the control path can be effectively removed/or atleast the code more made clean.

ControllerPublish and ControllerUnpublish will run only on linux nodes (but on any node). A proxy object redirection is necessary to perform this function on the replica node. Here volume gets attached and nfs server setup, along with sharedv4 service endpoint during setup phase. Controller can keep track of only portworx client nodes, but not external windows clients or outsiders. Not sure how its handled to keep refcounting correct.

For the client side, there would be NodeStageVolume followed by NodePublishVolume. 
NodeStageVolume as a first step, could setup the nfs client. 
NodePublishVolume could keep track of refcounts as necessary.

This is much cleaner abstraction than currently setup; and I would strongly vouch for the change.

Reply:
It could be argued either way whether adding NodeStageVolume would increase or decrease complexity. It doesn't make sense to add in this PR as it's a no-op in the current implementation anyway.

@sulakshm @pnookala-px as this is an initial alpha drop with limited testing, let's avoid sweeping design changes that could cause more instability for a small subset of windows use cases

Reply:
By adding the ControllerPublish/Unpublish or NodePublish/Unpublish all the way through the Driver and filter chain, we miss out on the functionality that we have added in Attach/Mount/Detach/Unmount in all the filters. Or you will have to add the same functionality in ControllerPublish & Unpublish in the filter implementation. I think this is more convoluted and duplication of effort.  We have a tried and battle tested path of Attach/Detach/Mount/Unmount and we should continue to use that. 

IMO, the controller/nodestage publish/unpublish is a contract between Kubernetes and the CSI driver. The CSI driver then needs to do what is required to achieve the publish/unpublish operation. And that doesn't mean adding this API all through the CSI driver's stack. So I am not bought on the clean code argument here.

Reply:
>  I think this is more convoluted and duplication of effort.
Attach and Mount calls currently are overloaded for solving many different issues, which you should be woefully aware. In the long run, I assumed we would split the functionality here. 

> We have a tried and battle tested path of Attach/Detach/Mount/Unmount and we should continue to use that.

This is definitely true; but the suggestion to split based on current complexity is already made.

> IMO, the controller/nodestage publish/unpublish is a contract between Kubernetes and the CSI driver. The CSI driver then needs to do what is required to achieve the publish/unpublish operation. And that doesn't mean adding this API all through the CSI driver's stack.

The part about CSI contract is true, but the extension to porx is debatable. If my already posted comments on Attach/Mount being overloaded with complexity does not make a convincing argument to split the function, then I have no more arguments to place.

In the meantime, I will sync up with @pnookala-px and see if changes can be limited first cut.